[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98320
* #98319
* #98300
* #98299
* __->__ #98250
* #98249
* #98221

This PR has `SHARD_GRAD_OP` (and `_HYBRID_SHARD_ZERO2`) skip `_use_sharded_views()` in the post-forward reshard since the strategy does not free the unsharded flat parameter and can preserve the unsharded views. This saves nontrivial CPU overhead both in the post-forward reshard (`_use_sharded_views()`) and the pre-backward unshard (`_use_unsharded_views()`).

<details>
<summary>(Before) Pre-backward hook: 4.356 ms</summary>

<img width="812" alt="Screenshot 2023-04-03 at 6 32 19 PM" src="https://user-images.githubusercontent.com/31054793/229641309-778cf1f9-4b5b-42ec-b2d8-0a1e6e7ce330.png">

</details>

<details>
<summary>(After) Pre-backward hook: 1.044 ms</summary>

![Screenshot 2023-04-04 at 9 05 53 AM](https://user-images.githubusercontent.com/31054793/229800917-9580ce6b-3721-469a-9212-f0cbfd8cbb52.png)

</details>

Commit:
[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`"




[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_use_orig_params.py:
```diff
@@ -448,6 +448,13 @@ def run_iter():
+            ShardingStrategy.SHARD_GRAD_OP,
+            ShardingStrategy._HYBRID_SHARD_ZERO2,
+        ):
+            # Skip the check since we do not expose the gradients after forward
```
Is this going to be an inconsistency? If so, do we document it somewhere?

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1055,7 +1063,7 @@ def pre_unshard(self) -> bool:
         """
         ret = False
-        if self._use_orig_params:
+        if self._use_orig_params and not self._skipped_use_sharded_views:
```
does this mean writeback feature gets disabled for zero2? can we just call use_sharded_views here?

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1517,12 +1530,24 @@ def _use_sharded_flat_param(self) -> None:
         flat_param.data = flat_param._local_shard  # type: ignore[attr-defined]
         if self._use_orig_params:
-            self._use_sharded_views()
+            in_forward = self._training_state == HandleTrainingState.FORWARD
```
does this work for activation checkpointing as well? or would we need AC to support a recomputing flag?

awgu replied:
I think it should work with AC for the cases we have been using FSDP <> AC. Previously, we already directly check `self._training_state == HandleTrainingState.FORWARD` to tell if we are in forward, regardless of recomputing or not. I will see if the unit tests fail though.

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1517,12 +1530,24 @@ def _use_sharded_flat_param(self) -> None:
+            if (
+                in_forward
+                and self._sharding_strategy
+                not in NO_RESHARD_AFTER_FORWARD_HANDLE_STRATEGIES
```
so we also skip calling use sharded grad views in this PR?

awgu replied:
Yes, I think we have to unless we want to use the `.data` hack to bypass the shape check since otherwise the parameters are unsharded while the gradients are sharded. This is a niche use case anyway since a gradient must be accumulated (either actually accumulated or `zero_grad(set_to_none=False)`).

rohan-varma reviewed (approved):
LGTM, thanks! 

If we're concerned about blast radius here, feel free to gate w an env var before landing.

Commit:
Update on "[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`"


(Before) Pre-backward hook: 4.356 ms
<img width="812" alt="Screenshot 2023-04-03 at 6 32 19 PM" src="https://user-images.githubusercontent.com/31054793/229641309-778cf1f9-4b5b-42ec-b2d8-0a1e6e7ce330.png">


(After) Pre-backward hook: 0.483 ms
<img width="1025" alt="Screenshot 2023-04-03 at 6 32 25 PM" src="https://user-images.githubusercontent.com/31054793/229641301-971d3c60-a4f1-4561-bb33-7aa07c42d0bd.png">

The "after" value might increase slightly if I add back the `_writeback_orig_params()` check.


[ghstack-poisoned]

Commit:
Update on "[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`"


(Before) Pre-backward hook: 4.356 ms
<img width="812" alt="Screenshot 2023-04-03 at 6 32 19 PM" src="https://user-images.githubusercontent.com/31054793/229641309-778cf1f9-4b5b-42ec-b2d8-0a1e6e7ce330.png">


(After) Pre-backward hook: 0.483 ms
<img width="1025" alt="Screenshot 2023-04-03 at 6 32 25 PM" src="https://user-images.githubusercontent.com/31054793/229641301-971d3c60-a4f1-4561-bb33-7aa07c42d0bd.png">

The "after" value might increase slightly if I add back the `_writeback_orig_params()` check.


[ghstack-poisoned]

Commit:
Update on "[FSDP] Skip `_use_sharded_views()` for `SHARD_GRAD_OP`"


(Before) Pre-backward hook: 4.356 ms
<img width="812" alt="Screenshot 2023-04-03 at 6 32 19 PM" src="https://user-images.githubusercontent.com/31054793/229641309-778cf1f9-4b5b-42ec-b2d8-0a1e6e7ce330.png">


(After) Pre-backward hook: 0.483 ms
<img width="1025" alt="Screenshot 2023-04-03 at 6 32 25 PM" src="https://user-images.githubusercontent.com/31054793/229641301-971d3c60-a4f1-4561-bb33-7aa07c42d0bd.png">

The "after" value might increase slightly if I add back the `_writeback_orig_params()` check.


[ghstack-poisoned]

awgu commented:
@pytorchbot merge
